### ISSUE TITLE
debugger doesn't step over lvalue subs

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13099]
Migrated from [rt.perl.org#118839](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118839) (status was 'resolved')

Searchable as RT118839$

**Comment by p5pRT** [comment]
### From perl-diddler@tlinx.org

#### Created by perl-diddler@tlinx.org



The perl debugger doesn't skip over lvalue subs but enters into
them\.

\-\-\-\-\-sample\-\-\-\-

\#\!/usr/bin/perl \-w

use strict; 
\{    package dat;
&nbsp;    our @&#8203;fields=qw\(my\_var\);

&nbsp;    sub new \{ my $p=shift; my $c=ref $p||$p;
&nbsp;                        bless $p=\{\}\, $c unless ref $p; $p \}

&nbsp;    sub my\_var \(;$\) :lvalue \{
&nbsp;        my $p=shift;
&nbsp;        if \(@&#8203;\_\) \{ $p\->\{my\_var\} = $\_\[0\] \}
&nbsp;        else \{ return $p\->\{my\_var\} \}
&nbsp;    \}

1;
\}
package main;
my $p = dat\->new;

$p\->my\_var = 23;

printf "my\_var = %s\\n"\, $p\->my\_var;

\-\-\-\-\-\-\-\-\-end sample\-\-\-\-

If you try to use the perl debugger to skip over the
"$p\->my\_var = 23" statement\, instead of stepping to
the following printf\, you instead find yourself at the
"my $p=shift;" statement in "sub my\_var"\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl 5.16.2 - Fri Feb 15 01:17:37 UTC 2013
It is being executed now by  Perl 5.16.2 - Fri Feb 15 01:12:05 UTC 2013.

Site configuration information for perl 5.16.2:

Configured by abuild at Fri Feb 15 01:12:05 UTC 2013.

Summary of my perl5 (revision 5 version 16 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=3.4.6-2.10-default, archname=x86_64-linux-thread-multi
    uname='linux build34 3.4.6-2.10-default #1 smp thu jul 26 09:36:26 utc 2012 (641c197) x86_64 x86_64 x86_64 gnulinux '
    config_args='-ds -e -Dprefix=/usr -Dvendorprefix=/usr -Dinstallusrbinperl -Dusethreads -Di_db -Di_dbm -Di_ndbm -Di_gdbm -Dd_dbm_open -Duseshrplib=true -Doptimize=-fmessage-length=0 -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -g -Wall -pipe -Accflags=-DPERL_USE_SAFE_PUTENV -Dotherlibdirs=/usr/lib/perl5/site_perl'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DPERL_USE_SAFE_PUTENV -fno-strict-aliasing -pipe -fstack-protector -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-fmessage-length=0 -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -g -Wall -pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DPERL_USE_SAFE_PUTENV -fno-strict-aliasing -pipe -fstack-protector'
    ccversion='', gccversion='4.7.2 20130108 [gcc-4_7-branch revision 195012]', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib64 -fstack-protector'
    libpth=/lib64 /usr/lib64 /usr/local/lib64
    libs=-lm -ldl -lcrypt -lpthread
    perllibs=-lm -ldl -lcrypt -lpthread
    libc=/lib64/libc-2.17.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.17'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.16.2/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib64 -fstack-protector'

Locally applied patches:
    


@INC for perl 5.16.2:
    /home/law/bin/lib
    /usr/lib/perl5/site_perl/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/site_perl/5.16.2
    /usr/lib/perl5/vendor_perl/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.16.2
    /usr/lib/perl5/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/5.16.2
    /usr/lib/perl5/site_perl/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/site_perl/5.16.2
    /usr/lib/perl5/site_perl
    .


Environment for perl 5.16.2:
    HOME=/home/law
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_CTYPE=en_US.UTF-8
    LD_LIBRARY_PATH=/usr/lib64/mpi/gcc/openmpi/lib64
    LOGDIR (unset)
    PATH=.:/home/law/bin/lib:/sbin:/usr/local/sbin:/usr/lib64/mpi/gcc/openmpi/bin:/home/law/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/X11R6/bin:/usr/games:/opt/kde3/bin:/usr/lib/mit/bin:/usr/lib/mit/sbin:/usr/lib/qt3/bin:/opt/dell/srvadmin/bin:/usr/sbin:/etc/local/func_lib:/home/law/lib
    PERL5OPT=-Mutf8 -CSA -I/home/law/bin/lib
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rjbs

\* Linda Walsh \<perlbug\-followup@&#8203;perl\.org> \[2013\-07\-09T23&#8203;:06&#8203;:19\]
> The perl debugger doesn't skip over lvalue subs but enters into
> them\.

Confirmed\.  Woah\, nice bug\.

I can confirm this on 5\.10\.1\, 5\.12\.4\, 5\.14\.2\, 5\.16\.0\, 5\.18\.0\.

It does \*not\* occur on 5\.8\.8\.

A debugger test for this might help get things moving\, if you feel like looking
at the debugger tests which are\, of late\, not \*too\* hard to produce\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ikegami

Replicated using 5\.18\.0&#8203;:

$ perl \-dE'
&nbsp;   use strict;
&nbsp;   use warnings;

&nbsp;   sub f :lvalue \{
&nbsp;      say "bar";
&nbsp;      $\_
&nbsp;   \}

&nbsp;   say "foo";
&nbsp;   f\(\) = 42;
&nbsp;   say "baz";
'

Loading DB routines from perl5db\.pl version 1\.39\_09
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:10\)&#8203;:     say "foo";
&nbsp;  DB\<1> n
foo
main&#8203;::\(\-e&#8203;:11\)&#8203;:     f\(\) = 42;
&nbsp;  DB\<1> n
main&#8203;::f\(\-e&#8203;:6\)&#8203;:        say "bar";    \<\-\-\- Unexpected
&nbsp;  DB\<1> n
bar
main&#8203;::f\(\-e&#8203;:7\)&#8203;:        $\_
&nbsp;  DB\<1> n
main&#8203;::\(\-e&#8203;:12\)&#8203;:     say "baz";
&nbsp;  DB\<1> n
baz
Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<1>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jul 10 06&#8203;:22&#8203;:45 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Linda Walsh \<perlbug\-followup@&#8203;perl\.org> \[2013\-07\-09T23&#8203;:06&#8203;:19\]
> > The perl debugger doesn't skip over lvalue subs but enters into
> > them\.
> 
> Confirmed\.  Woah\, nice bug\.
> 
> I can confirm this on 5\.10\.1\, 5\.12\.4\, 5\.14\.2\, 5\.16\.0\, 5\.18\.0\.
> 
> It does \*not\* occur on 5\.8\.8\.

It does not occur in 5\.8 because there was no mechanism back then for
handling lvalue subs specially \(via DB&#8203;::lsub\)\, so assignment to an
lvalue sub was completely broken under the debugger\.

Perl 5\.10 introduced DB&#8203;::lsub\.  The implementation in perl5db\.pl
contains these lines at the end of the sub&#8203;:

&nbsp;    \# Pop the single\-step value back off the stack\.
&nbsp;    $single |= $stack\[ $stack\_depth\-\- \];

&nbsp;    \# call the original lvalue sub\.
&nbsp;    &$sub;

The regular DB&#8203;::sub does this&#8203;:

&nbsp;        \{
&nbsp;            no strict 'refs';
&nbsp;            @&#8203;ret = &$sub;
&nbsp;        \}

&nbsp;        \# Pop the single\-step value back off the stack\.
&nbsp;        $single |= $stack\[ $stack\_depth\-\- \];

Notice how $single \(i\.e\.\, $DB&#8203;::single\) is modified before and after the
sub call\, respectively\.  The order is different\.

There are two types of lvalue list context for lvalue subs\.  \(foo\)=3
will allow sub foo&#8203;:lvalue\{@&#8203;a\} to return the array itself\.  \\\(foo\) and
bar\(foo\) will flatten the array\.

So there is no way in pure Perl to capture the return value and have it
returned to the caller unchanged\.  \(Maybe lvalue subs should be made
experimental again\. :\-\)

I have yet to come up with a solution\.  Even if I come up with a crazy
workaround\, we might want to consider an improved wantarray\(\) or
replacement\, and perhaps a way to force assignment context \(to prevent
flattening\) while capturing the values instead of actually doing an
assignment\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jul 13 02&#8203;:01&#8203;:29 2013\, sprout wrote&#8203;:
> I have yet to come up with a solution\.

I have just come across this nifty Perl feature called ‘local’\.

I used it to fix this bug in commit bf2614180a\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From perl-diddler@tlinx.org

On Sat Jul 13 11&#8203;:51&#8203;:22 2013\, sprout wrote&#8203;:
> On Sat Jul 13 02&#8203;:01&#8203;:29 2013\, sprout wrote&#8203;:
> > I have yet to come up with a solution\.
> 
> I have just come across this nifty Perl feature called ‘local’\.
> 
> I used it to fix this bug in commit bf2614180a\.
> 

Um\.\.\. that's nice\.\.\. I don't suppose you could be a bit more forthcoming
by documenting the solution/workaround here with the bug\, before closing
it as solved\.\.\.?   I don't know how to look up commit xxyz\.

;\-/

**Comment by p5pRT** [comment]
### From @mauke

On Sat Jul 13 14&#8203;:24&#8203;:29 2013\, LAWalsh wrote&#8203;:
> On Sat Jul 13 11&#8203;:51&#8203;:22 2013\, sprout wrote&#8203;:
> > On Sat Jul 13 02&#8203;:01&#8203;:29 2013\, sprout wrote&#8203;:
> > > I have yet to come up with a solution\.
> > 
> > I have just come across this nifty Perl feature called ‘local’\.
> > 
> > I used it to fix this bug in commit bf2614180a\.
> > 
> 
> Um\.\.\. that's nice\.\.\. I don't suppose you could be a bit more forthcoming
> by documenting the solution/workaround here with the bug\, before closing
> it as solved\.\.\.?   I don't know how to look up commit xxyz\.

http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/bf2614180a
> ;\-/

**Comment by p5pRT** [comment]
### From perl-diddler@tlinx.org

On Sat Jul 13 15&#8203;:19&#8203;:29 2013\, mauke\- wrote&#8203;:
> On Sat Jul 13 14&#8203;:24&#8203;:29 2013\, LAWalsh wrote&#8203;:
> > On Sat Jul 13 11&#8203;:51&#8203;:22 2013\, sprout wrote&#8203;:
> > > On Sat Jul 13 02&#8203;:01&#8203;:29 2013\, sprout wrote&#8203;:
> > > > I have yet to come up with a solution\.
> > > 
> > > I have just come across this nifty Perl feature called ‘local’\.
> > > 
> > > I used it to fix this bug in commit bf2614180a\.
> > > 
> > 
> > Um\.\.\. that's nice\.\.\. I don't suppose you could be a bit more forthcoming
> > by documenting the solution/workaround here with the bug\, before closing
> > it as solved\.\.\.?   I don't know how to look up commit xxyz\.
> 
> http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/bf2614180a
> > ;\-/


---
Thanks\!\.\.\. I wanted to apply it to my local copy\, since it
was a tad annoying\.  \(making the equivalent change in 5\.16\.2 worked out
just fine for my submitted test case\)\.\.\.
